core/GlobalState.cc

```diff
@@ -49,11 +50,11 @@ const int Symbols::MAX_PROC_ARITY;
 
 GlobalState::GlobalState(shared_ptr<ErrorQueue> errorQueue, shared_ptr<absl::Mutex> epochMutex,
                          shared_ptr<atomic<u4>> currentlyProcessingLSPEpoch, shared_ptr<atomic<u4>> lspEpochInvalidator,
-                         shared_ptr<atomic<u4>> lastCommittedLSPEpoch)
+                         shared_ptr<atomic<u4>> lastCommittedLSPEpoch, shared_ptr<shared_ptr<lsp::Task>> preemptTask)
```

Comment:
I don't currently understand why would it be a constructor argument, could you give more context?

Reply:
Also, do we expect Tasks to be immutable/threadsfe? The current code will share pointers between multiple global states(and thus threads)

Reply:
also, should it be `atomic<shared_ptr<lsp::Task>>`?

Reply:
> Also, do we expect Tasks to be immutable/threadsfe? The current code will share pointers between multiple global states(and thus threads)

Tasks themselves aren't threadsafe to run, but updating the `shared_ptr` atomically is threadsafe. I share the `shared_ptr` between GlobalStates so that we can update the task on `initialGS` in the message processing thread and it'll propagate to `finalGS` in the typechecking thread. It's similar to how initialGS and finalGS both share epoch counters.

> I don't currently understand why would it be a constructor argument, could you give more context?

I need to ensure that all GlobalStates from the same `initialGS` have the same `preemptTask` shared pointer.

> also, should it be atomic<shared_ptr<lsp::Task>>?

I recall having trouble using this formulation. Instead, I'm using the `shared_ptr` specializations of `atomic_load`/`atomic_store`:

https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic

Reply:
>. Tasks themselves aren't threadsafe to run

in the current api nothing a single task from being invoked from multiple threads concurrently.